### PR TITLE
Automatically enable the Strict Transport Security (HSTS) header

### PR DESCRIPTION
**Pull request by Toflar**
This will automatically send the `Strict-Transport-Security` header for all Contao requests (front end and back end). Imho, this always makes sense for `https` requests (for non-https, the header is irrelevant anyway).

I decided to enable it by default, I think this is the sane way to go. If you want to disable it, you can easily do that:

```yaml
contao:
    security:
        hsts: false # or enable: false
```

The TTL is configured to 1 year by default. You can also change this in case you want to:

```yaml
contao:
    security:
        hsts:
            ttl: 3600
```

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

**Comment by asaage** [comment]
I have this already configured via the webserver (it's an option in Plesk).
Wonder if it can be disabled or overridden by contao then - i dont't think so.
From experience: it can be a real pita if for some reason your cert is not valid (had to take care of that first and then restart my machine because the browser cached the cert or the failed hsts attempt in a funny way) there is also a way to exempt certain domains from that in the browser - but it's not easy to find.
Let's see how this works out.

**Comment by aschempp** [comment]
Not sure if this should be based on the current request security, or rather on the root page setting?

**Comment by Toflar** [comment]
Root page setting would be wrong because the header only makes sense when it is an https request.

**Comment by Toflar** [comment]
And it must also work in the backend.

**Comment by aschempp** [comment]
So if I'm on `http://` and the server would send a STS header, it would not be effective for the browser?

Also thinking about this:

> Should it be necessary to disable Strict Transport Security, setting the max-age to 0 (over an https connection) will immediately expire the Strict-Transport-Security header, allowing access via http.
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security

Should we send that `max-age=0` on a root page if root page is set to `http://` ? So if I switch the security setting in the root page, I would immediately be allowed to use http on that domain. Otherwise we would end up in an endless redirect, because Contao would redirect to http:// and the browser would enforce https:// ?

**Comment by Toflar** [comment]
> So if I'm on `http://` and the server would send a STS header, it would not be effective for the browser?

Jep, it would be ignored. Sending it over HTTP is even against [the specification](https://datatracker.ietf.org/doc/html/rfc6797):

> An HSTS Host MUST NOT include the STS header field in HTTP responses conveyed over non-secure transport.

> Should we send that `max-age=0` on a root page if root page is set to `http://` ? So if I switch the security setting in the root page, I would immediately be allowed to use http on that domain. Otherwise we would end up in an endless redirect, because Contao would redirect to http:// and the browser would enforce https:// ?

Even if that were possible, it wouldn't work for the backend. You'll receive the typical "potential security issue" screen of the browser (although the specification leaves that up to the browser vendors but if I think that would make most sense) and to get rid of it, you need to clear the cache.

**Comment by de-es** [comment]
Wenn ich den Code richtig interpretiere, tritt die Funktionalität dann in Konkurrenz zur selben aus dem Nelmio Security Bundle. D.h. wenn ich diese nutzen möchte, um ggf. noch HSTS für Subdomains zu steuern, müsste ich sie explizit abschalten, damit ich keine doppelten Header erhalte? Oder könntest Du die Einstellung für die Subdomains hier noch ergänzen, dann wäre die Konfiguration in Nelmio, zumindest für meinen Use Case, obsolet.

**Comment by Toflar** [comment]
I explicitly don't want subdomains to be present, we should not mess with those. And Nelmio will win now (https://github.com/contao/contao/pull/6553/commits/e5fff2824c77849d6b652a9055251df296875cf9).

**Comment by aschempp** [comment]
Enabling this by default can lead to issues having the front end reachable. If the backend is accessed with HTTPS but the root page is set to HTTP, Contao would send a redirect to HTTP and the browser would enforce HTTPS.

This can be solved with this PR: https://github.com/Toflar/contao/pull/14

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php [hunk 1]
```diff
@@ -509,4 +510,20 @@ private function getComposerPublicDir(string $projectDir): string|null
...
     }
+
+    private function handleSecurityConfig(array $config, ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('contao.listener.transport_security_header')) {
+            return;
+        }
+
+        if (false === $config['security']['hsts']['enabled']) {
+            $container->removeDefinition('contao.listener.transport_security_header');
```
Comment (aschempp) [inline core-bundle/src/DependencyInjection/ContaoCoreExtension.php:521]: Should we also disable the header on the `RedirectController` if it is disabled in the config?

Comment (Toflar) [reply core-bundle/src/DependencyInjection/ContaoCoreExtension.php:521]: I don't think that would be correct. Imho, you would always want that behavior.

File: core-bundle/config/listener.yaml [hunk 2]
```diff
@@ -566,6 +566,14 @@ services:
         tags:
             - kernel.event_subscriber
 
+    contao.listener.transport_security_header:
+        class: Contao\CoreBundle\EventListener\TransportSecurityHeaderListener
+        arguments:
+            - '@contao.routing.scope_matcher'
+            - ~
```
Comment (aschempp) [inline core-bundle/config/listener.yaml]: This is an invalid value for the service. I would suggest to not add this and make the service constructor property by setting a default value. The bundle config could then be nullable to apply the default value, or we could just leave it as-is.

File: core-bundle/src/EventListener/TransportSecurityHeaderListener.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly int $ttl,
+    ) {
+    }
+
+    public function __invoke(ResponseEvent $event): void
+    {
+        if (
+            !$this->scopeMatcher->isContaoMainRequest($event)
+            || !$event->getRequest()->isSecure()
+            || !$event->getRequest()->isMethodSafe()
```
Comment (aschempp) [inline core-bundle/src/EventListener/TransportSecurityHeaderListener.php]: didn't you want to remove this check, so we also set the header on POST requests for example?

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/config/listener.yaml (hunk 2 above)
Comment (Toflar) [reply core-bundle/config/listener.yaml]: Done.

File: core-bundle/src/EventListener/TransportSecurityHeaderListener.php (hunk 3 above)
Comment (Toflar) [reply core-bundle/src/EventListener/TransportSecurityHeaderListener.php]: Done.

**Review by aschempp — DISMISSED** [review]

File: core-bundle/config/listener.yaml [hunk 4]
```diff
@@ -566,6 +566,13 @@ services:
         tags:
             - kernel.event_subscriber
 
+    contao.listener.transport_security_header:
+        class: Contao\CoreBundle\EventListener\TransportSecurityHeaderListener
+        arguments:
+            - '@contao.routing.scope_matcher'
+        tags:
+            - kernel.event_listener
```
Comment (fritzmg) [inline core-bundle/config/listener.yaml:574]: `autoconfigure` is enabled for our event listeners, so if you prefer you could omit this and instead use the `#[AsEventListener]` PHP attribute.

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/config/listener.yaml (hunk 4 above)
Comment (Toflar) [reply core-bundle/config/listener.yaml:574]: We don't use that in any listener yet - I guess we should keep it unified.

File: core-bundle/tests/Controller/RedirectControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            ->willReturn($response)
+        ;
+
+        $request = Request::create('https://localhost/');
+
+        $controller = new RedirectController($inner);
+        $controller->urlRedirectAction($request, '/foo/bar');
+    }
+
+    public function testDoesNotAddTheHeaderIfRootPageUsesSSL(): void
```
Comment (leofeyer) [inline core-bundle/tests/Controller/RedirectControllerTest.php:101]: Isn‘t that the case where we want to add the header?

Comment (Toflar) [reply core-bundle/tests/Controller/RedirectControllerTest.php:101]: Ah, no that's actually correct. "The header" in this case refers to `Strict-Transport-Security': max-age=0` which is **unsetting** any potentially previously cached value. So if you use SSL you **do not** want the header that **unsets**. You only want the header if it should unset. 😊 
The one that actually adds it is not the `RedirectController` but the `TransportSecurityHeaderListener`. 
So yes, it is correct.

Comment (leofeyer) [reply core-bundle/tests/Controller/RedirectControllerTest.php:101]: Ah, I see. Thanks for the explanation.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**